### PR TITLE
HPCC-27646 Create theme editor

### PR DESCRIPTION
**Pull request by jeclrsg**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/src/src-react/hooks/theme.ts [hunk 1]
```diff
@@ -1,11 +1,88 @@
...
+
+    React.useEffect(() => {
+        if (themeDark === "dark") {
+            setTextColor(getShade(getColorFromString(textColor), Shade.Shade1).str);
+            setBackgroundColor(getBackgroundShade(getColorFromString(textColor), Shade.Shade8).str);
+        } else {
+            setTextColor(getShade(getColorFromString(textColor), Shade.Shade8).str);
+            setBackgroundColor(getBackgroundShade(getColorFromString("white"), Shade.Unshaded, true).str);
+        }
+    }, [setBackgroundColor, setTextColor, themeDark]);
```
Comment (GordonSmith) [inline esp/src/src-react/hooks/theme.ts:62]: Lint issues

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: esp/src/src-react/components/ThemeEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    color,
+    onColorChange,
+    label
+}) => {
+    const [pickerColor, setPickerColor] = React.useState(color);
+    const _colorPickerRef = React.useRef(null);
+    const [editingColorStr, setEditingColorStr] = React.useState(pickerColor?.str ?? "");
+    const [isColorPickerVisible, setIsColorPickerVisible] = React.useState<boolean>(false);
+
+    const _onTextFieldValueChange = (ev: any, newValue: string | undefined) => {
```
Comment (GordonSmith) [inline esp/src/src-react/components/ThemeEditor.tsx:41]: Should be a React.callback?

File: esp/src/src-react/hooks/theme.ts [hunk 3]
```diff
@@ -1,11 +1,89 @@
...
+    React.useEffect(() => {
+        if (themeDark === "dark") {
+            setTextColor(getShade(getColorFromString(textColor), Shade.Shade1).str);
+            setBackgroundColor(getBackgroundShade(getColorFromString(textColor), Shade.Shade8).str);
+        } else {
+            setTextColor(getShade(getColorFromString(textColor), Shade.Shade8).str);
+            setBackgroundColor(getBackgroundShade(getColorFromString("white"), Shade.Unshaded, true).str);
+        }
+        // excluding exhaustive-deps lint. effect shouldn't execute when textColor changes.
+    }, [setBackgroundColor, setTextColor, themeDark]); // eslint-disable-line react-hooks/exhaustive-deps
```
Comment (GordonSmith) [inline esp/src/src-react/hooks/theme.ts:63]: So the user never gets to set the actual text color?

**Review by GordonSmith — CHANGES_REQUESTED** [review]

**Comment by jeclrsg** [comment]
@GordonSmith I rebased to 8.8.x & fixed some merge conflicts in .../hooks/theme.ts. The color selection controls still work as expected. It doesn't update the v9 theme however. Should I update this PR to do so, or just worry about it whenever we start using any v9 components?

**Review by GordonSmith — APPROVED** [review]
@jeclrsg - I suspect this is going to need a global get/setter with broadcast capability?  Happy to pull and add if needed.

**Comment by ghalliday** [comment]
@jeclrsg @GordonSmith this seems to have stalled/need rebasing.

**Comment by ghalliday** [comment]
@jeclrsg @GordonSmith either needs rebasing, or closing because it is stale.

**Pull request closed without merging**